File: solutions/cpp/difference-of-squares/2/difference_of_squares.cpp
#include "difference_of_squares.h"
#include <numeric>
#include <algorithm>
#include <vector>

using namespace std;

namespace difference_of_squares
{

    uint64_t square_of_sum(uint32_t n)
    {
        vector<uint32_t> sequence(n);
        iota(begin(sequence), end(sequence), 1);

        auto sum = accumulate(begin(sequence), end(sequence), 0);

        return sum * sum;
    }

    uint64_t sum_of_squares(uint32_t n)
    {
        vector<uint32_t> sequence(n);
        iota(begin(sequence), end(sequence), 1);

        transform(
            begin(sequence), end(sequence), begin(sequence),
            [](auto value) { return value * value; });

        return accumulate(begin(sequence), end(sequence), 0);
    }

    uint64_t difference(uint32_t n)
    {
        return square_of_sum(n) - sum_of_squares(n);
    }

} // namespace difference_of_squares

File: solutions/fortran/difference-of-squares/2/difference_of_squares.f90
module difference_of_squares
  implicit none
contains

  pure integer function square_of_sum(n)
    integer, intent(in) :: n
    integer :: i
    
    square_of_sum = 0
    do i = 1,n
      square_of_sum = square_of_sum + i
    end do
    square_of_sum = square_of_sum ** 2
  end function square_of_sum

  pure integer function sum_of_squares(n)
    integer, intent(in) :: n
    integer :: i
    
    sum_of_squares = 0
    do i = 1,n
      sum_of_squares = sum_of_squares + i ** 2
    end do
  end function sum_of_squares

  pure integer function difference(n)
    integer, intent(in) :: n
    difference = square_of_sum(n) - sum_of_squares(n)
  end function difference

end module difference_of_squares
